curp/src/server/storage/wal/pipeline.rs

```diff
@@ -0,0 +1,170 @@
+}
+
+impl FilePipeline {
+    /// Creates a new `FilePipeline`
+    #[allow(clippy::integer_arithmetic)] // Introduced by tokio::select! macro
+    pub(super) fn new(dir: PathBuf, file_size: u64) -> io::Result<Self> {
+        Self::clean_up(&dir)?;
+
+        let (file_tx, file_rx) = flume::bounded(1);
+        let stop_event = Event::new();
+        let dir_c = dir.clone();
+
+        let handle = tokio::spawn(async move {
+            let mut file_count = 0;
+            loop {
+                let file = Self::alloc(&dir_c, file_size, file_count)?;
+                file_count += 1;
+                if let Err(e) = file_tx.send_async(file).await {
+                    // The receiver is already dropped, stop this task
+                    break;
+                }
+            }
+            Ok(())
+        });
+
+        Ok(Self {
+            dir,
+            file_size,
+            file_stream: file_rx.into_stream(),
+            handle,
+            stopped: false,
+        })
+    }
+
+    /// Stops the pipeline
+    pub(super) fn stop(&mut self) {
+        self.handle.abort();
+        self.stopped = true;
+        if let Err(e) = Self::clean_up(&self.dir) {
+            error!("failed to clean up pipeline files: {e}");
+        }
+    }
+
+    /// Allocates a a new tempfile
+    fn alloc(dir: impl AsRef<Path>, file_size: u64, file_count: usize) -> io::Result<LockedFile> {
+        let fpath = PathBuf::from(dir.as_ref()).join(format!("{file_count}{TEMP_FILE_EXT}"));
+        let mut locked_file = LockedFile::open_rw(fpath)?;
+        locked_file.preallocate(file_size)?;
+        Ok(locked_file)
+    }
+
+    /// Cleans up all unused tempfiles
+    fn clean_up(dir: &PathBuf) -> io::Result<()> {
+        for result in std::fs::read_dir(dir)? {
+            let file = result?;
+            if let Some(filename) = file.file_name().to_str() {
+                if filename.ends_with(TEMP_FILE_EXT) {
+                    std::fs::remove_file(file.path())?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Drop for FilePipeline {
+    fn drop(&mut self) {
+        self.stop();
+    }
+}
+
+impl Stream for FilePipeline {
+    type Item = io::Result<LockedFile>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        if self.handle.is_finished() {
+            if let Poll::Ready(result) = self.handle.poll_unpin(cx) {
+                match result {
+                    Ok(Err(e)) => {
+                        return Poll::Ready(Some(Err(e)));
+                    }
+                    Err(e) => {
+                        return Poll::Ready(Some(Err(e.into())));
+                    }
+                    Ok(Ok(_)) => return Poll::Ready(None),
+                }
+            }
+            return Poll::Ready(None);
+        }
+
+        if self.stopped {
+            return Poll::Ready(None);
+        }
```

Comment:
We can merge this piece of code. The `drop` method will invoke the `stop` method, aborting the handle and setting the stop flag to true. Therefore, the stop flag must be true when the `is_finished` returns true. If the `is_finished` returns false, the stop flag is still true. We can bank on this stop flag to cancel the stream.